### PR TITLE
Faster Reward Model Training with Sequential Packing

### PR DESCRIPTION
**Pull request by lomty**
### Summary
This PR refactors the reward model training pipeline to use **sequential packing** (`[context | chosen | rejected]`) instead of separate forward passes for chosen and rejected responses. This approach improves training efficiency, reduces memory overhead, and adds proper support for sequence parallelism.

#### 1. **Sequential Packing Architecture** (`turbo_alignment/dataset/pair_preferences/collators.py`)
- **New `PairPreferenceDataCollator`**: Implements sequential packing by concatenating context, chosen, and rejected segments into a single sequence
- **Custom 4D Attention Masks**: Enforces isolation between chosen and rejected segments while maintaining causal attention
- **Symmetric Position IDs**: Rejected segment mirrors chosen segment positions for fair comparison

Key methods:
- `_process_example()`: Packs `[context | chosen | rejected]` into single sequence
- `_get_attn_mask()`: Creates 4D causal masks with chosen/rejected isolation
- `_get_position_ids()`: Computes symmetric position IDs for fair reward comparison

#### 2. **Dataset Refactoring** (`turbo_alignment/dataset/pair_preferences/pair_preference.py`)
- **Moved batching logic from dataset to collator**: Dataset now returns individual context/chosen/rejected tokens
- **New `SplitChatDataset`**: Variant of `ChatDataset` that splits context and answer at encoding time
- **Simplified data flow**: Dataset handles tokenization, collator handles batching and packing
- **Fixed tensor size mismatches**: Proper padding ensures chosen/rejected have compatible shapes

Changes:
- `PairPreferenceDataset` now uses `SplitChatDataset` instead of `TrainChatDataset`
- Returns `inputs_context`, `inputs_chosen`, `inputs_rejected` instead of `inputs_w`, `inputs_l`
- Removed concatenation logic (moved to collator)

#### 3. **Chat Dataset Enhancement** (`turbo_alignment/dataset/chat/chat.py`)
- **Added `split_context_answer` parameter**: Enables context/answer splitting in `_encode()`
- **Bot message tracking**: Tracks where bot messages start to enable proper splitting
- **New `SplitChatDataset` class**: Dedicated class for split encoding
- **Improved `_truncate_and_merge()`**: Returns bot message start positions for splitting

Features:
- `bot_message_starts` list tracks bot message positions
- `context_ids` and `answer_ids` tensors for split output
- Backward compatible with existing chat dataset usage

#### 4. **Trainer Refactoring** (`turbo_alignment/trainers/rm.py`)
- **Removed `concatenated_forward()`**: No longer needed with sequential packing
- **Updated `compute_loss()`**: Single forward pass for all segments
- **Direct logit extraction**: Uses `model.model()` + `model.score()` for classification logits
- **Sequence parallel support**: Properly handles padding and chunking for sequence parallelism
- **Reward extraction**: Uses `chosen_indices` and `rejected_indices` from collator

Key improvements:
- Single forward pass instead of two separate passes
- Proper position_ids handling for sequence parallel
- Direct reward extraction from packed sequences
- Enhanced documentation

### Technical Details

**Sequential Packing Format:**
```
[context_0, context_1, ..., context_N | chosen_0, chosen_1, ..., chosen_M | rejected_0, rejected_1, ..., rejected_K]
```

**Attention Mask Strategy:**
- Context: Visible to both chosen and rejected
- Chosen: Causal attention, invisible to rejected
- Rejected: Causal attention, invisible to chosen

**Position IDs:**
- Context: `[0, 1, 2, ..., N-1]`
- Chosen: `[N, N+1, ..., N+M-1]`
- Rejected: `[N, N+1, ..., N+K-1]` (mirrors chosen positions)

### Benefits

1. **Efficiency**: Single forward pass instead of two separate passes
2. **Memory**: Reduced memory overhead from batching

### CODE REVIEW

File: turbo_alignment/modeling/qwen3.py [hunk 1]
```diff
@@ -120,12 +145,15 @@ def forward(
...
+        # gather_idx = 2
 
-        query_states = _SeqAllToAll().apply(spg, query_states, scatter_idx, gather_idx)
-        key_states = _SeqAllToAll().apply(spg, key_states, scatter_idx, gather_idx)
-        value_states = _SeqAllToAll().apply(spg, value_states, scatter_idx, gather_idx)
+        # query_states = _SeqAllToAll().apply(spg, query_states, scatter_idx, gather_idx)
+        # key_states = _SeqAllToAll().apply(spg, key_states, scatter_idx, gather_idx)
+        # value_states = _SeqAllToAll().apply(spg, value_states, scatter_idx, gather_idx)
+
+        key_states = GatherSeq.apply(key_states, spg)
```
Comment (pavelgein) [inline turbo_alignment/modeling/qwen3.py]: According to the Ulessys [paper](https://arxiv.org/pdf/2309.14509), Figure 2, you have to split different heads across  different GPUs, and gather all sequence on each GPU, not just gather

Comment (lomty) [reply turbo_alignment/modeling/qwen3.py]: Changed in the last version

**Review by pavelgein — APPROVED** [review]

**Review by alekseymalakhov11 — APPROVED** [review]

**Comment by mponty** [comment]
all tests passed
